---
title: "Identifying interaction workflow"
author: "Andrea Zermeño Díaz"
date: "`r Sys.Date()`"
output: html_document
---

Workflow of how to identify interaction patterns between ocean microbial phylogenetic groups and biosynthetic gene clusters (BGCs).

```{r libraries, warning=FALSE, message=FALSE}
library(tidyverse)
library(dplyr)
```

```{r load_data}
setwd("~/")
meta_mags <- read.csv("metadata.csv")
meta_bgcs <- read.csv("bgcs_metadata.csv")

# We are grouping by mOTUS so we need to omit the NAs / For other categories skip this 
meta_mags <- meta_mags %>% filter(!is.na(mOTUs.Species.Cluster))
meta_bgcs <- meta_bgcs %>%
  filter(Genome %in% meta_mags$Genome)
```

Count the number of Metagenome Assembled Genomes (MAGs) per site and the microbial phylogenetic group, for this try we will classify groups by mOTUs. Also for group the BGCs we will use the gene cluster families (GCFs).

```{r count_mags}
# number of mOTUs
#motus <- meta_mags %>% 
#  distinct(mOTUs.Species.Cluster)

# Count the number of MAGs per site and mOTUs
motus_by_sites <- meta_mags %>%
  count(station, mOTUs.Species.Cluster) %>%
  pivot_wider(names_from = mOTUs.Species.Cluster, values_from = n, values_fill = 0)
```

Add the locations to the BGCs table based on the MAGs from where they were found and count the number of BGCs per site and GCF.

```{r count_bgcs}
# bind the column station to each BGC using the column Genome ass a reference to match it 
bgcs_sites <- meta_bgcs %>%
  left_join(meta_mags %>% select(Genome, station), by = "Genome")

# number of GCFs 
#gcfs <- bgcs_sites %>% 
#  distinct(gcf)

# count the number of BGCs per site and GCF
gcfs_by_sites <- bgcs_sites %>%
  count(station, gcf) %>%
  pivot_wider(names_from = gcf, values_from = n, values_fill = 0) 
```

Now we need bind each column of the `gcfs_by_sites` table with each column of the `motus_by_sytes` table, so we can obtain 5,138 x 3,934 tables (20,212,892)  of 3 columns (stations, mOTUs and GCFs) and 256 rows dictated by station. This will help us to identify "co-exclusion" and "co-occurrence" patterns.


```{r}
# lista donde se van a guardar las tablas de patrones
co_occur <- data.frame(mags = character(), bgcs = character(), num_stations = integer(), stringsAsFactors = FALSE)

co_exclu <- data.frame(mags = character(), bgcs = character(), num_stations = integer(), stringsAsFactors = FALSE)

for (col1 in colnames(motus_by_sites)) {
  if (col1 == "station") next  # Ignoramos la columna station
  
  for (col2 in colnames(gcfs_by_sites)) {
    if (col2 == "station") next  # Ignoramos la columna station

    # select the station column and the current column for both tables
    temp1 <- motus_by_sites[, c("station", col1), drop = FALSE]
    temp2 <- gcfs_by_sites[, c("station", col2), drop = FALSE]
    
    temp3 <- full_join(temp1, temp2, by = "station")
    temp3[[col1]] <- ifelse(is.na(temp3[[col1]]), 0, temp3[[col1]])
    temp3[[col2]] <- ifelse(is.na(temp3[[col2]]), 0, temp3[[col2]])
    
    temp3 <- temp3[!(temp3[[col1]] == 0 & temp3[[col2]] == 0), ]
    if (nrow(temp3) == 0) next
    
    if (all((temp3[[col1]] > 0) == (temp3[[col2]] > 0))) { # co-ocurrencia
      co_occur <- rbind(co_occur, data.frame(mags= col1, bgcs= col2, num_stations= nrow(temp3)))
    }
    
    if (all((temp3[[col1]] == 0 & temp3[[col2]] > 0) | (temp3[[col1]] > 0 & temp3[[col2]] == 0))) {    # co-exclusion
      co_exclu <- rbind(co_exclu, data.frame(mags= col1, bgcs= col2, num_stations= nrow(temp3)))
    }
  }
}

# funcion que extraiga las columnas y haga el join para observar la tabla
# bgc que aparezca en muchas co exclusiones 
# graficar donde ocurre un patron en el mapa 
```


```{r}
top_10_co_exclu <- co_exclu[order(-co_exclu$num_stations), ][1:10, ]
print(top_10_co_exclu)
```


```{r funcition}
# Definir la función
recreate_table <- function(mag, bgc, motus_by_sites, gcfs_by_sites) {

  table1 <- motus_by_sites[, c("station", mag), drop = FALSE]
  table2 <- gcfs_by_sites[, c("station", bgc), drop = FALSE]
  
  table_comb <- full_join(table1, table2, by = "station")
  
  table_comb[[mag]] <- ifelse(is.na(table_comb[[mag]]), 0, table_comb[[mag]])
  table_comb[[bgc]] <- ifelse(is.na(table_comb[[bgc]]), 0, table_comb[[bgc]])
  
  # Eliminar filas donde ambos valores sean 0
  table_comb <- table_comb[!(table_comb[[mag]] == 0 & table_comb[[bgc]] == 0), ]
  
  return(table_comb)
}

# Ejemplo de uso de la función:
example <- recreate_table("gom_000200", "gcf_1033", motus_by_sites, gcfs_by_sites)
print(example)

```

#####################################################

```{r try3, eval=FALSE}

motus_cols <- setdiff(colnames(motus_by_sites), "station")
gcfs_cols <- setdiff(colnames(gcfs_by_sites), "station")

# Generate all the possible combinations
combinaciones <- map2(
  rep(motus_cols, each = length(gcfs_cols)), 
  rep(gcfs_cols, times = length(motus_cols)),
  ~ motus_by_sites %>%
    select(station, .x) %>%
    full_join(gcfs_by_sites %>% select(station, .y), by = "station") %>%
    rename_with(~ c("station", .x, .y), everything())
)
```

```{r try1, eval=FALSE}
combinaciones <- list()

contador <- 1
for (col1 in colnames(motus_by_sites)) {
  for (col2 in colnames(gcfs_by_sites)) {
    combinaciones[[contador]] <- data.frame(motus_by_sites[[col1]], gcfs_by_sites[[col2]])
    names(combinaciones[[contador]]) <- c(col1, col2)
    contador <- contador + 1
  }
}

#

# names(combinaciones) <- paste(rep(colnames(tabla1), each = ncol(tabla2)),
#                              rep(colnames(tabla2), times = ncol(tabla1)),
#                              sep = "_")

```

```{r check, eval=FALSE}
# 1138 mOTUs of only 1 MAG representative 
meta_mags %>%
  group_by(mOTUs.Species.Cluster) %>%                     
  summarise(n = n()) %>%                 
  filter(n == 1)  
# 2564 gcfs of only 1 BGC representative 
bgcs_sites %>%
  group_by(gcf) %>%                     
  summarise(n = n()) %>%                 
  filter(n == 1)  

names(which(table(meta_mags$mOTUs.Species.Cluster) == 1))



setdiff(motus_by_sites$station, gcfs_by_sites$station)
setdiff(meta_mags$station, bgcs_sites$station)
```




